File: solutions/c/scrabble-score/1/scrabble_score.c
#include "scrabble_score.h"

unsigned int score(const char *word) {
    const unsigned int POINTS[] = {1, 3, 3, 2, 1, 4, 2, 4, 1, 8, 5, 1, 3, 1, 1, 3, 10, 1, 1, 1, 1, 4, 4, 8, 4, 10};
    unsigned int total = 0;

    for (; *word != 0; word += 1) {
        const char letter = *word;
        unsigned int index = letter - (letter >= 'a' ? 'a' : 'A');
        total += POINTS[index];
    }

    return total;
}

File: solutions/ruby/scrabble-score/1/scrabble_score.rb
class Scrabble

  @@points_for = Proc.new do |letter|
    case letter.downcase
    when 'a', 'e', 'i', 'o', 'u', 'l', 'n', 'r', 's', 't' then 1
    when 'd', 'g'                                         then 2
    when 'b', 'c', 'm', 'p'                               then 3
    when 'f', 'h', 'v', 'w', 'y'                          then 4
    when 'k'                                              then 5
    when 'j', 'x'                                         then 8
    when 'q', 'z'                                         then 10
    else
      0
    end
  end

  def self.score(word)
    word.to_s.chars.map(&@@points_for).reduce(0, &:+)
  end

  def initialize(word)
    @word = word
  end

  def score
    Scrabble.score(@word)
  end

end
